From 493eaec7cff4d24452a689759d5d6e4f8eb6034b Mon Sep 17 00:00:00 2001
From: Paul Junod <donujp@gmail.com>
Date: Tue, 13 Oct 2026 06:17:44 -0400
Subject: [PATCH] frames/window-tracker: Use the GTK interface color scheme
 property

GTK 4.20 deprecated GtkSettings:gtk-application-prefer-dark-theme in
favour of GtkSettings:gtk-interface-color-scheme. Set the new property
from the desktop color scheme, keeping the existing mapping: dark when
the desktop prefers dark, light otherwise.

The enum and the property only exist since GTK 4.20, so raise the GTK 4
requirement from 4.14 to 4.20.

Link: TBD-MR
---
 meson.build                      | 2 +-
 src/frames/meta-window-tracker.c | 4 +++-
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/meson.build b/meson.build
index cecc3c7bc..3104892c6 100644
--- a/meson.build
+++ b/meson.build
@@ -20,7 +20,7 @@ glib_req = '>= 2.81.1'
 gi_req = '>= 0.9.5'
 graphene_req = '>= 1.10.2'
 gtk3_req = '>= 3.19.8'
-gtk4_req = '>= 4.14.0'
+gtk4_req = '>= 4.20.0'
 pango_req = '>= 1.46.0'
 cairo_req = '>= 1.10.0'
 pangocairo_req = '>= 1.20'
diff --git a/src/frames/meta-window-tracker.c b/src/frames/meta-window-tracker.c
index 3f08388b5..885ce62a1 100644
--- a/src/frames/meta-window-tracker.c
+++ b/src/frames/meta-window-tracker.c
@@ -102,7 +102,9 @@ update_color_scheme (MetaWindowTracker *window_tracker)
   is_dark = color_scheme == G_DESKTOP_COLOR_SCHEME_PREFER_DARK;
 
   g_object_set (gtk_settings_get_default (),
-                "gtk-application-prefer-dark-theme", is_dark,
+                "gtk-interface-color-scheme",
+                is_dark ? GTK_INTERFACE_COLOR_SCHEME_DARK
+                        : GTK_INTERFACE_COLOR_SCHEME_LIGHT,
                 NULL);
 }
 
